lasy/propagators/angular_spectrum_dfft_propagator.py

```diff
@@ -0,0 +1,189 @@
+        describing the refractive index along the frequency/wavelength axis of the
+        laser pulse, or a function of the wavelength (in meters).
+        Default value is n=1. to describe propagation in vacuum.
+    """
+
+    def __init__(self, omega0, dim, n=1.0):
+        super().__init__()
+        self.update(dim=dim, omega0=omega0, n=n)
+
+    def update(self, dim, omega0, n):
+        r"""
+        Initialize or update the propagator if needed.
+
+        Parameters
+        ----------
+        dim : string
+            Dimensionality of the array. Options are:
+            - ``'xyt'``: Laser pulse represented on a 3D Cartesian grid.
+            - ``'rt'`` : Laser pulse represented on a 2D cylindrical grid.
+
+        omega0 : float (in rad.s^-1)
+            The main frequency :math:`\omega_0`, which is defined by the laser
+            wavelength :math:`\lambda_0`, as :math:`\omega_0 = 2\pi c/\lambda_0`.
+
+        n : int, float, 1d array or callable
+            Refractive index of the medium in which to propagate the laser.
+            Can be either a single value if dispersive effects are ignored, a 1d array
+            describing the refractive index along the frequency/wavelength axis of the
+            laser pulse, or a function of the wavelength (in meters).
+            Default value is n=1. to describe propagation in vacuum.
+        """
+        assert isinstance(n, (int, float, np.ndarray)) or callable(n)
+        assert dim in ["rt", "xyt"]
+
+        self.n = n  # refractive index
+        self.omega0 = omega0
+        self.dim = dim
+
+    def propagate(self, distance, grid_in, dim=None, omega0=None, grid_out=None):
+        r"""
+        Propagates the laser field in z diration by a given distance using the angular spectrum method.
+
+        Parameters
+        ----------
+        distance : scalar
+            Distance by which the laser is propagated.
+
+        grid_in : Grid
+            Grid object containing the laser to propagate.
+
+        dim : string (optional)
+            Dimensionality of the array. Options are:
+            - ``'xyt'``: Laser pulse represented on a 3D Cartesian grid.
+            - ``'rt'`` : Laser pulse represented on a 2D cylindrical grid.
+
+        omega0 : float (optional)
+            The main frequency :math:`\omega_0` (in rad.s^-1), which is defined by the laser
+            wavelength :math:`\lambda_0`, as :math:`\omega_0 = 2\pi c/\lambda_0`.
+
+        grid_out : Grid object (optional)
+            Grid object on which the propagated laser pulse is defined.
+            Can be different from laser grid before propagation.
+
+        Returns
+        -------
+        Grid object with laser data after propagation.
+
+        """
+        dim = dim if dim is not None else self.dim
+        omega0 = omega0 if omega0 is not None else self.omega0
+        self.update(omega0=omega0, dim=dim, n=self.n)
```

Comment:
Could the checks be done within `self.update`?
```suggestion
        self.update(omega0=omega0, dim=dim, n=self.n)
```
The idea of `update` is described in the corresponding base class's docstring:
> Typically called at the beginning of the propagate function, to update the propagator itself if required.
If the update is costly, some machinery should be in place to do it only if required.
Some propagators need to be updated if the distance to propagate changes.
Others only need to be updated if the grid properties change, etc.

In your case, I don't think there's expensive operations to do, so `update` is not doing much but that's fine.

I am also not sure about the logic within `update`: would this be safer?
```py
self.dim = dim if dim is not None else self.dim
assert self.dim is not None
```